File: js/shared/cpu/z80/instructions/Z80Bitwise.js
/**
 * Project: EGGStation - Unified Multi-System Console Virtual Environment
 * Author: Enrique González Gutiérrez
 * File: js/shared/cpu/z80/instructions/Z80Bitwise.js
 * 
 * Role:
 * Domain Layer: Z80 CPU Bitwise Instructions.
 * This class encapsulates all Z80 CPU instructions designed for individual bit 
 * manipulation (BIT, SET, RES). The logic follows the Command Pattern, dynamically 
 * registering instruction closures against the primary CPU instruction decoders.
 * 
 * SOLID Principles Applied:
 * 1. Single Responsibility Principle (SRP): Exclusively responsible for single-bit 
 *    test, set, and clear operation definitions. It isolates dynamic table mapping 
 *    logic completely from standard execution loops.
 * 2. Interface Segregation Principle (ISP): It expects only a thin, unified 
 *    opcode mapping dictionary (registry) rather than depending on the full, 
 *    heavy execution context of the ZilogZ80 class.
 */

class Z80Bitwise {
    /**
     * Registers all Bitwise opcodes onto the provided CPU opcode maps.
     * @param {ZilogZ80} cpu - The CPU Orchestrator instance.
     * @param {Z80Registers} registers - The CPU Registers state object.
     * @param {Z80Alu} alu - The Arithmetic Logic Unit for flag processing.
     * @param {Object} registry - The categorized opcode mapping arrays.
     */
    static register(cpu, registers, alu, registry) {

        /**
         * Helper for displacement address computation used in index-relative 
         * addressing (e.g., SET 3,(IX+d)).
         * @param {number} indexValue - Base 16-bit index (IX or IY).
         * @returns {number} The absolute 16-bit memory offset.
         */
        const getDisplacement = (indexValue) => {
            const d = cpu.theMMU.readAddr(registers.pc + 2);
            // Sign-extend the 8-bit displacement value (-128 to 127)
            const incr = (d & 0x80) === 0x80 ? -0x80 + (d & 0x7F) : d;
            return (indexValue + incr) & 0xffff;
        };

        // ========================================================================
        // DYNAMIC BITWISE INSTRUCTION GENERATOR (BIT, RES, SET)
        // ========================================================================
        
        // Z80 Standard Register mapping sequence: B, C, D, E, H, L, (HL), A
        const regNames = ['B', 'C', 'D', 'E', 'H', 'L', '(HL)', 'A'];
        const regProps = ['b', 'c', 'd', 'e', 'h', 'l', null, 'a'];

        for (let bit = 0; bit < 8; bit++) {
            const bitMask = 1 << bit;
            const notBitMask = (~bitMask) & 0xff;

            for (let r = 0; r < 8; r++) {
                const isHL = (r === 6);
                const regName = regNames[r];
                const prop = regProps[r];

                // Calculate base opcodes for the CB prefix matrix
                const opBit = 0x40 + (bit * 8) + r;
                const opRes = 0x80 + (bit * 8) + r;
                const opSet = 0xC0 + (bit * 8) + r;

                // --- 1. STANDARD CB-PREFIXED OPERATIONS ---
                
                // BIT b, r
                registry.bitwise[opBit] = [() => {
                    const val = isHL ? cpu.theMMU.readAddr(registers.hl) : registers[prop];
                    alu.bit_8bit(registers, val, bitMask);
                    cpu.incPc(2);
                }, `BIT ${bit},${regName}`, isHL ? 12 : 8, 0, false];

                // RES b, r
                registry.bitwise[opRes] = [() => {
                    if (isHL) {
                        let val = cpu.theMMU.readAddr(registers.hl);
                        val &= notBitMask;
                        cpu.theMMU.writeAddr(registers.hl, val);
                    } else {
                        registers[prop] &= notBitMask;
                    }
                    cpu.incPc(2);
                }, `RES ${bit},${regName}`, isHL ? 15 : 8, 0, false];

                // SET b, r
                registry.bitwise[opSet] = [() => {
                    if (isHL) {
                        let val = cpu.theMMU.readAddr(registers.hl);
                        val |= bitMask;
                        cpu.theMMU.writeAddr(registers.hl, val);
                    } else {
                        registers[prop] |= bitMask;
                    }
                    cpu.incPc(2);
                }, `SET ${bit},${regName}`, isHL ? 15 : 8, 0, false];


                // --- 2. INDEXED DDCB / FDCB OPERATIONS (IX/IY overrides) ---
                // In the Z80 architecture, IX/IY index bitwise instructions occupy 
                // the exact same matrix slots as the (HL) pointer (Register index 6).
                if (isHL) {
                    
                    // BIT b, (IX+d) / (IY+d)
                    registry.bitwiseIX[opBit] = [() => {
                        const val = cpu.theMMU.readAddr(getDisplacement(registers.ix));
                        alu.bit_8bit(registers, val, bitMask);
                        cpu.incPc(4);
                    }, `BIT ${bit},(IX+%d)`, 20, 1, false];

                    registry.bitwiseIY[opBit] = [() => {
                        const val = cpu.theMMU.readAddr(getDisplacement(registers.iy));
                        alu.bit_8bit(registers, val, bitMask);
                        cpu.incPc(4);
                    }, `BIT ${bit},(IY+%d)`, 20, 1, false];

                    // RES b, (IX+d) / (IY+d)
                    registry.bitwiseIX[opRes] = [() => {
                        const addr = getDisplacement(registers.ix);
                        let val = cpu.theMMU.readAddr(addr);
                        val &= notBitMask;
                        cpu.theMMU.writeAddr(addr, val);
                        cpu.incPc(4);
                    }, `RES ${bit},(IX+%d)`, 23, 1, false];

                    registry.bitwiseIY[opRes] = [() => {
                        const addr = getDisplacement(registers.iy);
                        let val = cpu.theMMU.readAddr(addr);
                        val &= notBitMask;
                        cpu.theMMU.writeAddr(addr, val);
                        cpu.incPc(4);
                    }, `RES ${bit},(IY+%d)`, 23, 1, false];

                    // SET b, (IX+d) / (IY+d)
                    registry.bitwiseIX[opSet] = [() => {
                        const addr = getDisplacement(registers.ix);
                        let val = cpu.theMMU.readAddr(addr);
                        val |= bitMask;
                        cpu.theMMU.writeAddr(addr, val);
                        cpu.incPc(4);
                    }, `SET ${bit},(IX+%d)`, 23, 1, false];

                    registry.bitwiseIY[opSet] = [() => {
                        const addr = getDisplacement(registers.iy);
                        let val = cpu.theMMU.readAddr(addr);
                        val |= bitMask;
                        cpu.theMMU.writeAddr(addr, val);
                        cpu.incPc(4);
                    }, `SET ${bit},(IY+%d)`, 23, 1, false];
                }
            }
        }
    }
}

File: src/domain/cpu/z80/instructions/Z80Bitwise.cpp
/**
 * Project: EGGStation - Unified Multi-System Console Virtual Environment
 * Author: Enrique González Gutiérrez
 * File: src/domain/cpu/z80/instructions/Z80Bitwise.cpp
 * 
 * Domain Layer: Z80 CPU Bitwise Instructions Implementation
 */

#include "Z80Bitwise.h"

static inline uint16_t getDisplacement(ZilogZ80& cpu, uint16_t indexValue) {
    uint8_t d = cpu.mmu->readAddr(cpu.registers.pc + 2);
    int8_t incr = static_cast<int8_t>(d);
    return (indexValue + incr) & 0xFFFF;
}

// ========================================================================
// CORE BITWISE OPERATIONS TEMPLATES (Compile-time resolution)
// ========================================================================

template<int bit, int reg, bool isHL>
static void executeBit(ZilogZ80& c) {
    uint8_t val = 0;
    if (isHL) {
        val = c.mmu->readAddr(c.registers.getHL());
    } else {
        if (reg == 0)      val = c.registers.b;
        else if (reg == 1) val = c.registers.c;
        else if (reg == 2) val = c.registers.d;
        else if (reg == 3) val = c.registers.e;
        else if (reg == 4) val = c.registers.h;
        else if (reg == 5) val = c.registers.l;
        else if (reg == 7) val = c.registers.a;
    }
    c.alu.bit_8bit(c.registers, val, 1 << bit);
    c.incPc(2);
}

template<int bit, int reg, bool isHL>
static void executeRes(ZilogZ80& c) {
    uint8_t notBitMask = ~(1 << bit) & 0xFF;
    if (isHL) {
        uint8_t val = c.mmu->readAddr(c.registers.getHL());
        val &= notBitMask;
        c.mmu->writeAddr(c.registers.getHL(), val);
    } else {
        if (reg == 0)      c.registers.b &= notBitMask;
        else if (reg == 1) c.registers.c &= notBitMask;
        else if (reg == 2) c.registers.d &= notBitMask;
        else if (reg == 3) c.registers.e &= notBitMask;
        else if (reg == 4) c.registers.h &= notBitMask;
        else if (reg == 5) c.registers.l &= notBitMask;
        else if (reg == 7) c.registers.a &= notBitMask;
    }
    c.incPc(2);
}

template<int bit, int reg, bool isHL>
static void executeSet(ZilogZ80& c) {
    uint8_t bitMask = 1 << bit;
    if (isHL) {
        uint8_t val = c.mmu->readAddr(c.registers.getHL());
        val |= bitMask;
        c.mmu->writeAddr(c.registers.getHL(), val);
    } else {
        if (reg == 0)      c.registers.b |= bitMask;
        else if (reg == 1) c.registers.c |= bitMask;
        else if (reg == 2) c.registers.d |= bitMask;
        else if (reg == 3) c.registers.e |= bitMask;
        else if (reg == 4) c.registers.h |= bitMask;
        else if (reg == 5) c.registers.l |= bitMask;
        else if (reg == 7) c.registers.a |= bitMask;
    }
    c.incPc(2);
}

// ========================================================================
// INDEXED BITWISE OPERATIONS TEMPLATES (DDCB / FDCB)
// ========================================================================

template<int bit>
static void executeBitIX(ZilogZ80& c) {
    uint8_t val = c.mmu->readAddr(getDisplacement(c, c.registers.getIX()));
    c.alu.bit_8bit(c.registers, val, 1 << bit);
    c.incPc(4);
}

template<int bit>
static void executeBitIY(ZilogZ80& c) {
    uint8_t val = c.mmu->readAddr(getDisplacement(c, c.registers.getIY()));
    c.alu.bit_8bit(c.registers, val, 1 << bit);
    c.incPc(4);
}

template<int bit>
static void executeResIX(ZilogZ80& c) {
    uint16_t addr = getDisplacement(c, c.registers.getIX());
    uint8_t val = c.mmu->readAddr(addr);
    val &= ~(1 << bit) & 0xFF;
    c.mmu->writeAddr(addr, val);
    c.incPc(4);
}

template<int bit>
static void executeResIY(ZilogZ80& c) {
    uint16_t addr = getDisplacement(c, c.registers.getIY());
    uint8_t val = c.mmu->readAddr(addr);
    val &= ~(1 << bit) & 0xFF;
    c.mmu->writeAddr(addr, val);
    c.incPc(4);
}

template<int bit>
static void executeSetIX(ZilogZ80& c) {
    uint16_t addr = getDisplacement(c, c.registers.getIX());
    uint8_t val = c.mmu->readAddr(addr);
    val |= 1 << bit;
    c.mmu->writeAddr(addr, val);
    c.incPc(4);
}

template<int bit>
static void executeSetIY(ZilogZ80& c) {
    uint16_t addr = getDisplacement(c, c.registers.getIY());
    uint8_t val = c.mmu->readAddr(addr);
    val |= 1 << bit;
    c.mmu->writeAddr(addr, val);
    c.incPc(4);
}

// ========================================================================
// HIGH-PERFORMANCE STATIC INSTRUCTION BINDERS
// ========================================================================

template<int bit, int reg>
static inline void bindBitwise(Z80OpcodeRegistry& registry) {
    constexpr bool isHL = (reg == 6);
    constexpr uint8_t opBit = 0x40 + (bit * 8) + reg;
    constexpr uint8_t opRes = 0x80 + (bit * 8) + reg;
    constexpr uint8_t opSet = 0xC0 + (bit * 8) + reg;

    registry.bitwise[opBit] = { &executeBit<bit, reg, isHL>, static_cast<uint8_t>(isHL ? 12 : 8) };
    registry.bitwise[opRes] = { &executeRes<bit, reg, isHL>, static_cast<uint8_t>(isHL ? 15 : 8) };
    registry.bitwise[opSet] = { &executeSet<bit, reg, isHL>, static_cast<uint8_t>(isHL ? 15 : 8) };
}

template<int bit>
static inline void bindIndexedBitwise(Z80OpcodeRegistry& registry) {
    constexpr uint8_t opBit = 0x40 + (bit * 8) + 6;
    constexpr uint8_t opRes = 0x80 + (bit * 8) + 6;
    constexpr uint8_t opSet = 0xC0 + (bit * 8) + 6;

    registry.bitwiseIX[opBit] = { &executeBitIX<bit>, 20 };
    registry.bitwiseIY[opBit] = { &executeBitIY<bit>, 20 };
    registry.bitwiseIX[opRes] = { &executeResIX<bit>, 23 };
    registry.bitwiseIY[opRes] = { &executeResIY<bit>, 23 };
    registry.bitwiseIX[opSet] = { &executeSetIX<bit>, 23 };
    registry.bitwiseIY[opSet] = { &executeSetIY<bit>, 23 };
}

// ========================================================================
// CORE REGISTRATION PIPELINE
// ========================================================================

void Z80Bitwise::registerInstructions(ZilogZ80& cpu, Z80OpcodeRegistry& registry) {

    // --- 1. Bit 0 ---
    bindBitwise<0, 0>(registry); bindBitwise<0, 1>(registry); bindBitwise<0, 2>(registry); bindBitwise<0, 3>(registry);
    bindBitwise<0, 4>(registry); bindBitwise<0, 5>(registry); bindBitwise<0, 6>(registry); bindBitwise<0, 7>(registry);
    bindIndexedBitwise<0>(registry);

    // --- 2. Bit 1 ---
    bindBitwise<1, 0>(registry); bindBitwise<1, 1>(registry); bindBitwise<1, 2>(registry); bindBitwise<1, 3>(registry);
    bindBitwise<1, 4>(registry); bindBitwise<1, 5>(registry); bindBitwise<1, 6>(registry); bindBitwise<1, 7>(registry);
    bindIndexedBitwise<1>(registry);

    // --- 3. Bit 2 ---
    bindBitwise<2, 0>(registry); bindBitwise<2, 1>(registry); bindBitwise<2, 2>(registry); bindBitwise<2, 3>(registry);
    bindBitwise<2, 4>(registry); bindBitwise<2, 5>(registry); bindBitwise<2, 6>(registry); bindBitwise<2, 7>(registry);
    bindIndexedBitwise<2>(registry);

    // --- 4. Bit 3 ---
    bindBitwise<3, 0>(registry); bindBitwise<3, 1>(registry); bindBitwise<3, 2>(registry); bindBitwise<3, 3>(registry);
    bindBitwise<3, 4>(registry); bindBitwise<3, 5>(registry); bindBitwise<3, 6>(registry); bindBitwise<3, 7>(registry);
    bindIndexedBitwise<3>(registry);

    // --- 5. Bit 4 ---
    bindBitwise<4, 0>(registry); bindBitwise<4, 1>(registry); bindBitwise<4, 2>(registry); bindBitwise<4, 3>(registry);
    bindBitwise<4, 4>(registry); bindBitwise<4, 5>(registry); bindBitwise<4, 6>(registry); bindBitwise<4, 7>(registry);
    bindIndexedBitwise<4>(registry);

    // --- 6. Bit 5 ---
    bindBitwise<5, 0>(registry); bindBitwise<5, 1>(registry); bindBitwise<5, 2>(registry); bindBitwise<5, 3>(registry);
    bindBitwise<5, 4>(registry); bindBitwise<5, 5>(registry); bindBitwise<5, 6>(registry); bindBitwise<5, 7>(registry);
    bindIndexedBitwise<5>(registry);

    // --- 7. Bit 6 ---
    bindBitwise<6, 0>(registry); bindBitwise<6, 1>(registry); bindBitwise<6, 2>(registry); bindBitwise<6, 3>(registry);
    bindBitwise<6, 4>(registry); bindBitwise<6, 5>(registry); bindBitwise<6, 6>(registry); bindBitwise<6, 7>(registry);
    bindIndexedBitwise<6>(registry);

    // --- 8. Bit 7 ---
    bindBitwise<7, 0>(registry); bindBitwise<7, 1>(registry); bindBitwise<7, 2>(registry); bindBitwise<7, 3>(registry);
    bindBitwise<7, 4>(registry); bindBitwise<7, 5>(registry); bindBitwise<7, 6>(registry); bindBitwise<7, 7>(registry);
    bindIndexedBitwise<7>(registry);
}